crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+                let parent = self.stack.last().unwrap();
+
+                let section = Section {
+                    // HEADER_RE can't match without a match for group 'title'.
+                    title: caps.name("title").unwrap().into(),
+                    level: header_level,
+                    parent_id: Some(*parent),
+                };
+
+                if self.current_section_files.is_some() {
+                    return Err(anyhow::anyhow!(
+                        "Header '{}' not valid inside a test case; parent '{}' has code files.",
+                        section.title,
+                        self.sections[*parent].title,
+                    ));
+                }
+
+                let section_id = self.sections.push(section);
+                self.stack.push(section_id);
+
+                self.current_section_files = None;
+            } else if let Some(caps) = self.scan(&CODE_RE) {
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let mut config: FxHashMap<&'s str, &'s str> = FxHashMap::default();
+
+                if let Some(config_match) = caps.name("config") {
+                    for item in config_match.as_str().split_whitespace() {
+                        let mut parts = item.split('=');
+                        let key = parts.next().unwrap();
+                        let Some(val) = parts.next() else {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        };
+                        if parts.next().is_some() {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        }
+                        config.insert(key, val);
+                    }
+                }
+
+                let path = config.get("path").unwrap_or(&"test.py");
+
+                self.files.push(File {
+                    path,
+                    section: *parent,
+                    // CODE_RE can't match without matches for 'lang' and 'code'.
+                    lang: caps.name("lang").unwrap().into(),
+                    code: caps.name("code").unwrap().into(),
+                });
+
+                if let Some(current_files) = &mut self.current_section_files {
+                    if current_files.contains(*path) {
+                        let current = &self.sections[*self.stack.last().unwrap()];
+                        return Err(anyhow::anyhow!(
+                            "Test `{}` has duplicate files named `{path}`.",
+                            current.title
+                        ));
+                    }
+                    current_files.insert(*path);
+                } else {
+                    self.current_section_files = Some(FxHashSet::from_iter([*path]));
+                }
+            } else {
+                self.scan(&IGNORE_RE);
```

Comment:
Using a regex to find the end of a line seems overly complicated. It took me a while to figure out what's happening. 

I prefer calling `self.rest.lines()` and updating `self.rest` in place. 

What's even the case where we hit `IGNORE_RE` where it isn't the end? Shouldn't the other patterns match for as long as there are sections / code blocks and the `IGNORE_RE` only gets executed when there are no more sections/code blocks which means we're done and we can set `self.rest = ""`. Should we warn about any content coming after?

Reply:
I want to support arbitrary Markdown prose/markup interspersed in tests, as comments/description (even the sample `numbers.md` included in this PR uses this.) So we need to ignore whatever lines we find in between code blocks and headers.

Splitting by lines makes things a lot more complex because then I can't use a single multi-line regex to match an entire code block, I have to maintain my own state machine for code block parsing. That's possible of course, but I don't think "IGNORE_RE is confusing" is strong enough justification to spend the time on it.

Reply:
I mean just using `self.rest.lines()` in the case the other two regexes don't match. Although it's unclear to me if that's even necessary or if we can just move to the end because what's an example where the header and code block regex don't match, but they do match after ignoring.

... Or is it because you don't use the multiline flag and this indeed won't be necessary when you use the multiline flag?

Reply:
I feel like maybe you didn't read the first sentence of my comment :) 

> I want to support arbitrary Markdown prose/markup interspersed in tests, as comments/description (even the sample numbers.md included in this PR uses this.) So we need to ignore whatever lines we find in between code blocks and headers.

Reply:
I did... it's still unclear in which cases the first and second regex start matching again after they didn't, at least when using the regex multiline option.

Reply:
Oh! Maybe you're right that the combination of multi-line regex option and not anchoring the header/code patterns at start-of-string will be enough to remove the need for explicit ignoring entirely. I will experiment. (And if that doesn't work, I can still avoid using a regex for what is effectively find-next-newline.)

Reply:
As I commented above, multi-line option doesn't work (or at least would require significant changes to the parsing algorithm to make it work) because it's too greedy and will happily bypass things that shouldn't be bypassed. 

I switched to using `.find('\n')` and `.split_at` instead of `IGNORE_RE`.